### ISSUE TITLE
60 TPS ? (parity aura v1.11.11)

### ISSUE THREAD

**Issue by drandreaskrueger** [paritytech/parity-ethereum#9393]
I am benchmarking Ethereum based PoA chains, with my toolbox [chainhammer](https://gitlab.com/electronDLT/chainhammer).

My initial results for a dockerized network of **parity aura v1.11.8** nodes ... 

... leaves space for improvements :-)

Initial -unoptimized- run:

![chainreader/img/parity-poa-playground_run1_tps-bt-bs-gas_blks108-211.png](https://gitlab.com/electronDLT/chainhammer/raw/master/chainreader/img/parity-poa-playground_run1_tps-bt-bs-gas_blks108-211.png)

More details here: https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#benchmarking

### Please help ...

... by suggesting what we could try, to get this faster than 60 TPS. 

(Ideally approx 8 times faster, to beat [quorum IBFT](https://gitlab.com/electronDLT/chainhammer/blob/master/quorum-IBFT.md#result-400-tps-but-only-for-the-first-14k-tx).) 

Thanks a lot!  
Andreas

---

> I'm running:
>
> - **Which Parity version?**: v1.11.8
> - **Which operating system?**: Linux
> - **How installed?**: via docker
> - **Are you fully synchronized?**: not relevant
> - **Which network are you connected to?**: private chain
> - **Did you try to restart the node?**: yes
> - **actual behavior**   slow  
> - **expected behavior** faster
> - **steps to reproduce**: See [parity.md](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md).

**Comment by Tbaut** [comment]
There is definitely a lot of room for improvement :)
Can you share your setup: 
- how many authorities?
- how many nodes?
- are those well connected (same network) ?
- what is the hardware configuration of the nodes/authorities?
- how do you broadcast your Txs (do you sign them locally and send them to a node or do you let parity sign them)

**Comment by 5chdn** [comment]
What's the block gas limit and aura block time?

Please share config and chain spec.

**Comment by drandreaskrueger** [comment]
Fantastic, thanks for all the hints (and the [tweet](https://twitter.com/5chdn/status/1032749019179765760) ;-) )

Answers are all in [here](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md), probably mainly below [here](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#orbita-centerparity-poa-playground). 

The author/issue-answerer of that parity-poa-playground seems grumpy - so I am happy that [the parity-deploy.sh team](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#parity-deploy-docker-compose-generator-tool) is really responsive & helpful. Will try that again instead, next week.

Back in the office on Monday. Really looking forward to an optimized run. Have a good weekend, everyone.

**Comment by 5chdn** [comment]
> What's the block gas limit and aura block time?

Ok, I see, this is pretty much your answer ^

You have [three authorities with one second block time](https://github.com/orbita-center/parity-poa-playground/blob/master/parity/config/chain.json#L6-L11) and [a gas floor target of 6 billion](https://github.com/orbita-center/parity-poa-playground/blob/master/parity/config/authority.toml#L22).

This is a very good configuration to test TPS however, it does start with a lower block gas limit as it moves slowly up to the target. Did you consider running it for an extended period of time (hours, days) or simply modify the network configuration to start with a very high block gas limit, yet?

**Comment by drandreaskrueger** [comment]
> however, it does start with a lower block gas limit 

not sure about that. 

(1) 
Look at the bottom right diagram, it shows gasUsed and gasLimit   
and gas is *not* maxed out, in contrast to other runs (scroll down all the way [here](https://gitlab.com/electronDLT/chainhammer/blob/master/chainreader/blocksDB_analyze_quorum-istanbul.ipyn) in this early run - there the TPS was clearly limited by gasLimit).

Also, [gasLimit(https://github.com/orbita-center/parity-poa-playground/blob/master/parity/config/chain.json#L40) is set to be high, no?

But I might not understand all those parameters yet. Then sorry. What can I read?

Most importantly, my time is limited: Feel free to (simply run chainhammer yourself, or) send me any other configuration that you think will perform better. Does `authority.toml` & `chain.json` fully define that, or are there more settings files?

Thanks a lot.

**Comment by drandreaskrueger** [comment]
> --jsonrpc-server-threads maybe set to 4 or 8.

Great.
That looks promising. Earlier in my endeavour, I was surprised to see how little multithreading actually helped with parity (*) but this could explain that yes.

Is that a parity only setting, or can geth do that too?

**Comment by drandreaskrueger** [comment]
(*) Actually, back then it was the energywebfoundation "tobalaba" fork of parity. By the way, they left some [issues unanswered](https://gitlab.com/electronDLT/chainhammer/blob/master/tobalaba.md#issues), perhaps anyone of you guys has ideas; after all it is parity 1.8.0, right?

**Comment by 5chdn** [comment]
don't use the ewf client. parity ethereum now supports chain tobalaba

**Comment by drandreaskrueger** [comment]
> parity ethereum now supports chain tobalaba

Great.

I just lost a lot of time on benchmarking their outdated client then.  

And their chain was one big hickup, until they fixed that.  

Anything in their issues though, that might be useful to answer? Thanks!

**Comment by drandreaskrueger** [comment]
> This is a very good configuration to test TPS

Good. Still, if you know a better setup than that, I am happy to try that next week.


> however, it does start with a lower block gas limit 

not sure about that. 

(1) 
Look at the bottom right diagram [here](https://gitlab.com/electronDLT/chainhammer/blob/master/chainreader/img/parity-poa-playground_run1_tps-bt-bs-gas_blks108-211.png), it shows gasUsed and gasLimit   
and gas is *not* maxed out, in contrast to other runs (scroll down all the way [here](https://gitlab.com/electronDLT/chainhammer/blob/master/chainreader/blocksDB_analyze_quorum-istanbul.ipynb) in this early run - there the TPS was clearly limited by gasLimit).

(2)
Also, [gasLimit](https://github.com/orbita-center/parity-poa-playground/blob/master/parity/config/chain.json#L40) is set to be 0x165A0BC00 = 6 billion, no?

But I might not understand all those parameters yet. Then sorry. What can I read, which parameters are influencing TPS?

Most importantly, as my time is limited: 

>  simply modify the network configuration

Feel free to (simply run chainhammer yourself, or) send me any other configuration that you think will perform better. Does `authority.toml` & `chain.json` fully define that, or are there more settings files?

Thanks a lot.

**Comment by drandreaskrueger** [comment]
### new tests 
but

### no acceleration!

See https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run2 and below.

**63 TPS**

that is slow.

... new ideas please. Thanks.

**Comment by drandreaskrueger** [comment]
new run6 with some more parameters added, see description of the run in https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run6

--> **65 TPS**

**Comment by ddorgan** [comment]
Just replicating your setup now. But maybe a --gas-floor-target of something more realistic would be a good idea ... e.g. maybe 20m ... Also moving the stepDuration to about 3 to 4 seconds would be needed in a real life situation when not just benchmarking everything on one host.

**Comment by drandreaskrueger** [comment]
> Just replicating your setup now. 

Great, I am happy. Thanks a lot for your time, @ddorgan @5chdn and @Tbaut.  Let's find out how we can get `parity` faster - ideally as fast as `geth`, no?

And? Got it running? Which rates are you able to see?   Or: Need help?   

 

> --gas-floor-target of something more realistic would be a good idea ... e.g. maybe 20m ...   

Thanks. Explanation:

> --gas-floor-target Amount of gas per block to target when sealing a new block (default: 4700000).  
> https://hudsonjameson.com/2017-06-27-accounts-transactions-gas-ethereum/  

Please tell me about all those parameters which might be able to accelerate the current setup.  A warning: This whole benchmarking is admittedly not a "realistic" setup which rates to expect when running a network of nodes distributed on the planet. Internet bandwidth, ping times, etc. will always slow it down; I am looking at it more as an attempt to identify the current upper limit; any realistic setup will always be slower than what I have measured. However, if you want to, you can quite as well create a network with nodes on each continent, and then use chainhammer to benchmark that. So:

> Also moving the stepDuration to about 3 to 4 seconds

Yes.  

However, I have tested [most of the other systems](https://gitlab.com/electronDLT/chainhammer/blob/985180359c7f244013721bf482387838a27126e3/README.md#chronology) with a fast block rate too; my initial focus was on [quorum, and raft consensus](https://gitlab.com/electronDLT/chainhammer/blob/985180359c7f244013721bf482387838a27126e3/quorum.md#example-output) has sub-seconds block rate, and [quorum IBFT](https://gitlab.com/electronDLT/chainhammer/blob/985180359c7f244013721bf482387838a27126e3/quorum-IBFT.md#direct-rpc-call-instead-of-web3-call) runs without problems with 1 second block times - without the internet, of course.

Parity however is not even adhering to its own parameter, I suppose [this is a target blocktime of 2 seconds, right?](https://github.com/paritytech/parity-deploy/blob/master/config/spec/example.spec#L4-L7) - and [the run with 4 nodes then results in 4-8 seconds blocktime](https://gitlab.com/electronDLT/chainhammer/blob/985180359c7f244013721bf482387838a27126e3/parity.md#run6) !  So aren't we already at the blockspeed that you are suggesting?


These are the parameters that I have been adding to the standard https://github.com/paritytech/parity-deploy

```
--geth 
--jsonrpc-server-threads 100
--tx-queue-size 20000
--cache-size 4096
--tracing off
--gas-floor-target 20000000
--pruning fast
--tx-queue-mem-limit 0
--no-dapps
--no-secretstore-http
```
because I found them somewhere in an issue about speed.  Which of them are irrelevant?

**Comment by drandreaskrueger** [comment]
## 65 TPS

settings & log of run7: https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run7

results diagrams: https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#result-run-7

![https://gitlab.com/electronDLT/chainhammer/raw/master/chainreader/img/parity-aura_run7_tps-bt-bs-gas_blks3-90.png](https://gitlab.com/electronDLT/chainhammer/raw/master/chainreader/img/parity-aura_run7_tps-bt-bs-gas_blks3-90.png)

There is a new [README.md --> quickstart](https://gitlab.com/electronDLT/chainhammer/blob/master/README.md#quickstart) now ...

... so if you have any intution or knowledge how to accelerate this, please replicate my setup, and then start modifying the parameters of the network of parity nodes, e.g. with [parity-deploy](https://github.com/paritytech/parity-deploy) - until you get to better TPS rates. 

Then please alert us how you did it. Thanks.

**Comment by drandreaskrueger** [comment]
I found a new upper limit of **69 TPS**, but with [instantseal and only 1 node](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-8-instantseal).

**Comment by AyushyaChitransh** [comment]
In around June(don't exactly recall which parity version), we made some scripts to transfer ether from one account to another account. This script pushed transactions to  5 server setup(across different geographical regions) and we were able to achieve maximum TPS of 5001.

 - For block gas limit same as of mainnet(8000029), 3001 tx in one block
 - For increased block gas limit(upto 100 times of mainnet limit) 5001 tx in one block
 - For increased block gas limit(upto 10000 times of mainnet limit) 5001 tx in one block

These results were obtained from not one but multiple blocks.

## Environment:
 - StepDuration: 1, which ensures 1 block issuance per second
 - Number of Authorities: 5
 - Consensus Mechanism: Aura

Other than this, there were no more changes to `--jsonrpc-server-threads`

Most likely this low TPS that you are able to achieve is because of contract transactions.

## Cause:
Contract transactions are more heavy in terms of amount of gas used. So few transactions are able to completely fill up the block. Maybe increasing gas limit would allow you to get a higher TPS.

Regarding instantSeal: This may not be required in benchmarking any ethereum client, as it is not built to be decentralised (the core idea of blockchain). Even if one person was able to get a higher TPS on instantSeal, this blockchain would not be of much use to other willing to be a part of that network.

**Comment by 5chdn** [comment]
I don't have time to reproduce this myself right now @drandreaskrueger - but 65 sounds wrong by orders of magnitude. what @AyushyaChitransh reports sounds more realistic.

**Comment by drandreaskrueger** [comment]
> we were able to achieve maximum TPS of 5001.

Your TPS benchmarking sounds impressive. 

Please (someone) publish the exact commands - perhaps in a dockerized form, so that we can replicate it easily within a few minutes.  I am happy to [include it here in my repo](https://gitlab.com/electronDLT/chainhammer/blob/master/other-projects.md).

> Most likely this low TPS that you are able to achieve is because of contract transactions.

Yes to "lower" but no to "low" - because I am putting geth, quorum, etc through the exact same task. 

And: Simple money transfer is not relevant for our use case.

My storage `.set(x)` transaction takes [less than 27000 gas](https://gitlab.com/electronDLT/chainhammer/blob/master/chainreader/README.md#notes), which is at least 10 times your gas usage, right? (your 8000029/3001= ...) So if gas were the only parameter, I should be able to see 300-500 TPS - which I do with geth, but not with parity. Something else but the EVM must be slowing down parity.

I am always using the same transaction, on geth, quorum, parity, etc. - so the TPS values are comparable, right?

*I have never benchmarked simple money transfer, because that is not what we do.* Instead we need to **choose the fastest client for smart contract transactions**, and that is currently *quorum IBFT* (with  over 400 TPS) or *geth Clique* (with over 300 TPS), and they were measured in the exactly same way as I measured parity.  Please see [quorum-IBFT.md](https://gitlab.com/electronDLT/chainhammer/blob/master/quorum-IBFT.md) and [geth.md](https://gitlab.com/electronDLT/chainhammer/blob/master/geth.md). It is a pity, because for years we have always preferred *parity* and it would mean that we have to revise quite a bit of our inhouse code - but we simply cannot ignore a 6 times faster TPS.

@AyushyaChitransh please you repeat your benchmarking with a simplemost smart contract transaction - storing one value in a contract (or doing one multiplication and one addition - I still want to do that, but haven't had the time yet, see [TODO.md](https://gitlab.com/electronDLT/chainhammer/blob/master/TODO.md)). 

See [the call](https://gitlab.com/electronDLT/chainhammer/blob/master/send.py#L85) and [the contract](https://gitlab.com/electronDLT/chainhammer/blob/master/contract.sol)

>  Maybe increasing gas limit would allow you to get a higher TPS.

Done that, been there.

Please have a look at the bottom right diagram in each of my measurements, then you can see `gasLimit` and `gasUsed` per second.  When I see one touching the other, I know that I have to raise the gasLimit. Which I did. In the last runs I used 40 million gas as a limit. But no, the blocks were not maxed out.

Thanks for all your answers, but please spend some time looking at my stuff first, thanks.

> StepDuration: 1, which ensures 1 block issuance per second

No, it doesn't. I have set mine to [2 seconds](https://github.com/paritytech/parity-deploy/blob/master/config/spec/engine/aura#L5), but almost always it [ended up to be 4-8 seconds](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run7)


> across different geographical regions

That is a more realistic setup, to include the effects of the internet.  For now, I am benchmarking the client itself, and all my 4-7 nodes  are running on the same one machine, a 2016 desktop. But e.g. CPU is *not* maxing out, it stays around 50% during the whole benchmarking.

> Regarding instantSeal: This may not be required in benchmarking any ethereum client

I know. Of course. But it is the most simplistic thing I can ask parity to do, and then I have still seen less than 70 TPS.

Please **anyone** now replicate that setup of [run 8](https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-8-instantseal), with the now new [quickstart](https://gitlab.com/electronDLT/chainhammer/blob/master/README.md#quickstart) manual of chainhammer.



> I don't have time to reproduce this myself right now

I get that, @5chdn . We are all busy. And as you can see in my TODO.md list cited above, I also have some unfinished tasks with this. However, until you our someone else is disproving my findings, it already looks as if we might have our (yes, still preliminary) results: 

For our purposes `parity` is ~6 times slower than `geth` .



>  65 sounds wrong by orders of magnitude

Yes, and I am suprised about that myself.

The whole intention of all these interactions here is to get **anyone** who is more knowledgable about parity than me, to help me find the problem - and fix it. If your team is too small, what about employing more people? Or perhaps there is someone else out there, who would work for no pay?  Please help, thanks.

**Comment by drandreaskrueger** [comment]
# chainhammer
## How to replicate the results

```
# docker
# this is for Debian Linux, 
# if you run a different distro, google "install docker [distro name]"
sudo apt-get update 
sudo apt-get -y remove docker docker-engine docker.io 
sudo apt-get install -y apt-transport-https ca-certificates wget software-properties-common
wget https://download.docker.com/linux/debian/gpg 
sudo apt-key add gpg
rm gpg
echo "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | sudo tee -a /etc/apt/sources.list.d/docker.list
sudo apt-get update
sudo apt-cache policy docker-ce
sudo apt-get -y install docker-ce
sudo systemctl start docker

usermod -aG docker ${USER}
groups $USER
```

```
# parity-deploy
# for a dockerized parity environment
# this is instantseal, NOT a realistic network of nodes
# for 8 different setups see chainhammer-->parity.md
git clone https://github.com/paritytech/parity-deploy.git paritytech_parity-deploy
cd paritytech_parity-deploy
sudo ./clean.sh
./parity-deploy.sh --config dev --name instantseal --geth
docker-compose up
```

```
# solc
# someone should PLEASE create a Debian specific installation routine
# see https://solidity.readthedocs.io/en/latest/installing-solidity.html 
# and https://github.com/ethereum/solidity/releases
wget https://github.com/ethereum/solidity/releases/download/v0.4.24/solc-static-linux
chmod 755 solc-static-linux 
echo $PATH
sudo mv solc-static-linux /usr/local/bin/
sudo ln -s /usr/local/bin/solc-static-linux /usr/local/bin/solc
solc --version
Version: 0.4.24+commit.e67f0147.Linux.g++
```


```
# chainhammer & dependencies
git clone https://gitlab.com/electronDLT/chainhammer electronDLT_chainhammer
cd electronDLT_chainhammer/

pip install virtualenv
virtualenv -p python3 py3eth
source py3eth/bin/activate

python3 -m pip install --upgrade pip==9.0.3
pip3 install --upgrade py-solc==2.1.0 web3==4.3.0 web3[tester]==4.3.0 rlp==0.6.0 eth-testrpc==1.3.4 requests 
pip3 install pandas jupyter ipykernel matplotlib 
ipython kernel install --user --name="Python.3.py3eth"
```

```
# configure chainhammer
nano config.py

RPCaddress, RPCaddress2 = 'http://localhost:8545', 'http://localhost:8545'
ROUTE = "web3"
```

```
# test connection
touch account-passphrase.txt
./deploy.py 
```

```
# start the chainhammer viewer
./tps.py
```

new terminal


```
# same virtualenv
cd electronDLT_chainhammer/
source py3eth/bin/activate

# start the chainhammer send routine
./deploy.py notest; ./send.py 
```

new terminal

```
# check that the transactions are actually successfully executed:

geth attach http://localhost:8545

> web3.eth.getTransaction(web3.eth.getBlock(50)["transactions"][0])
{
  gas: 90000, ...
}

> web3.eth.getTransactionReceipt(web3.eth.getBlock(50)["transactions"][0])
{ 
  gasUsed: 26691,
  status: "0x1", ...
}
> 
```

Now about "not having the time" - these 2.5 hours today happened on my FREE DAY. I must convince them now that I can take those hours off again.

**Comment by drandreaskrueger** [comment]
So, I have invested one more workday, to make it even easier quicker faster simpler for you @AyushyaChitransh @5chdn @Tbaut and @ddorgan ... to have a 8 minutes look at the problem:

# **a readymade Amazon AMI** 
!!!

with *everything* pre-installed. Only look at [reproduce.md#readymade-amazon-ami](https://gitlab.com/electronDLT/chainhammer/blob/master/reproduce.md#readymade-amazon-ami) if you need to know what I did. Other than that, simply ...

**... use my AMI:**

* In the Public Images, [search for "chainhammer"](https://eu-west-2.console.aws.amazon.com/ec2/v2/home?region=eu-west-2#Images:visibility=public-images;search=chainhammer;sort=name)
* Right click ... Launch
* Step 2: Choose an Instance Type --> at least `t2.small` ! otherwise you probably run out of memory
* Step 4: Add Storage --> `12 GB` !
* security group, choose an existing ssh keypair e.g. `AndreasKeypairAWS.pem` (obviously, your keypair instead)
* launch, wait

simplify the `ssh` access, by adding this (with *your* IP, obviously) to your local machine's
```
nano ~/.ssh/config
```
```
Host chainhammer
  Hostname ec2-35-178-181-232.eu-west-2.compute.amazonaws.com
  StrictHostKeyChecking no
  User admin
  IdentityFile ~/.ssh/AndreasKeypairAWS.pem
```
now it becomes this simple to connect:
```
ssh chainhammer
```


### parity
```
ssh chainhammer

cd ~/paritytech_parity-deploy
sudo ./clean.sh
./parity-deploy.sh --config dev --name instantseal --geth
docker-compose up
```
For more complex setups than instantseal, see [parity.md](parity.md).

If you want to end this ... 'Ctrl-c' and:

```
docker-compose down -v
sudo ./clean.sh
```

### geth

```
ssh chainhammer

cd ~/drandreaskrueger_geth-dev/
docker-compose up
```

If you want to end this ... 'Ctrl-c' and:

```
docker-compose down -v
```

### chainhammer: test connection
... and create some local files
```
ssh chainhammer
cd electronDLT_chainhammer && source py3eth/bin/activate

./deploy.py
```
If there are connection problems, probably need to configure the correct ports in [config.py](config.py):
```
nano config.py
```


### chainhammer: watcher
```
./tps.py
```

### chainhammer: send transactions
```
ssh chainhammer
cd electronDLT_chainhammer && source py3eth/bin/activate

./deploy.py notest; ./send.py threaded2 23
```

Thanks for your time.

Any idea how to accelerate parity ... much appreciated!

**Comment by tnpxu** [comment]
#8960 just found this , can u try with this settings

**Comment by drandreaskrueger** [comment]
## Final verdict:

| hardware  	| node type 	| #nodes  	| peak TPS_av 	| final TPS_av 	|
|-----------	|-----------	|--------	|--------		|--------------	|
| | | |    	|         	|          |
| t2.large 	| parity    	| 4      	    	| 53.5        	|  52.9        |
| t2.xlarge 	| parity    	| 4      	 	| 56.5        	|  56.1        |
| t2.2xlarge 	| parity    	| 4      	 	| 57.6        	|  57.6        |
| | | |    	|         	|          |
| t2.2xlarge 	| geth      	| 3+1    	   	| 421.6       	| 400.0        	|
| t2.xlarge 	| geth      	| 3+1    	  	| 386.1       	| 321.5        	|
| t2.large 	    | geth      	| 3+1      	| 170.7       	| 169.4        	|
| t2.small 	    | geth      	| 3+1    	| 96.8       	| 96.5        	|

Giving up now.

**Comment by ddorgan** [comment]
@drandreaskrueger fyi a good reference from Cody Born @ Microsoft .... 

https://twitter.com/codyborn/status/1040081548135948288

Maybe  less taxing contract calls though ...

**Comment by drandreaskrueger** [comment]
Thanks. Scroll down to the "Development" --> "Programmatically interacting with a smart contract" chapter on https://docs.microsoft.com/en-us/azure/blockchain-workbench/ethereum-poa-deployment which is linked in that tweet.

What I see there:

(1)  
his smart contract is roughly as simple as [mine](https://gitlab.com/electronDLT/chainhammer/blob/master/contract.sol) - I store an int, he stores a string.

(2)   
he is signing the transaction himself:  
```
   tx.sign(privateKey);
   var raw = '0x' + tx.serialize().toString('hex');
   web3.eth.sendRawTransaction(raw, function (txErr, transactionHash) { ...
```

(3)  
He does not tell us how he benchmarked. I could not yet find his "using a simple perf tool I put together.". So his results are simply not reproducible. 

If you find all of his code, feel free to point me to it.

(4)  
It is very probably not an issue of Azure vs Amazon. He sees 400 TPS on Azure, and I see less than 60 TPS on Amazon AWS. But [the exact same Amazon AWS machine which lets `parity aura` speed up to sleepy 58 TPS ... makes `geth clique` run much faster at 400 TPS](https://gitlab.com/electronDLT/chainhammer/blob/master/reproduce.md#results).


**What else can you read out of his article?**

Asking because of all the above points ... only (3) is seems really relevant.  --> If that is the underlying reason for `geth` handling transactions 6 times faster than `parity` - then I suggest that you **have a look at your parity transaction signing code** - perhaps that slows down parity so dramatically?

And again: If you don't believe my numbers, reproduce them yourself. You need not more than a few minutes if you start here:  https://gitlab.com/electronDLT/chainhammer/blob/master/reproduce.md#readymade-amazon-ami

Until you come up with faster `parity --settings" ... I have my results: `geth` is at least 5 times faster than parity.

**Comment by drandreaskrueger** [comment]
Of all the above points ... only (2) is seems really relevant.  

I am using [contract.functions.set( x=arg ).transact(txParameters)](https://gitlab.com/electronDLT/chainhammer/blob/69407bab63ca56fb62bae736c9cf3e6a7c32c6c0/send.py#L85) via web3, or [eth_sendTransaction](https://gitlab.com/electronDLT/chainhammer/blob/69407bab63ca56fb62bae736c9cf3e6a7c32c6c0/send.py#L167) via direct RPC call.  

So I let the node do the transaction signing. How else would I do it, as 23 multi-threading threads are hammering at the same node, and it is not clear in which order the transactions are arriving, I would not know which nonce to use in each of those transactions - right?

In contrast he seems to sign the *contract-deployment* transaction himself, and then does a `web3.eth.sendRawTransaction`.  Perhaps he does the same when hammering? We don't know, as I could not find his benchmarking code yet.

 --> If that is the underlying reason for `geth` handling transactions 6 times faster than `parity` - then I suggest that you parity people **have a long look at your parity transaction signing code** - perhaps that slows down parity so dramatically?  Perhaps *compare your code with geth`s and see what they do differently.*

**Comment by drandreaskrueger** [comment]
## Giving up now
Again and again and again: If you don't believe my numbers, reproduce them yourself. 

I have put an immense amount of work into this, with the single goal that *you* need not more than a few minutes to reproduce:  https://gitlab.com/electronDLT/chainhammer/blob/master/reproduce.md#readymade-amazon-ami 

Be cooperative for once, and run it, and tell me what you see. Please.

And then make a suggestion how to tune `parity`.

I really really really want to see it fast. Why? We at electron.org.uk have invested time into, and built our code around `parity`. It would be a pain to switch to `geth` now. But:

Until you come up with faster `parity --settings` ... I must believe my sad results: 

`geth` is at least 5 times faster than parity.


:-(

**Comment by 5chdn** [comment]
I'm sorry to hear that.

**Comment by drandreaskrueger** [comment]
what?

**Comment by drandreaskrueger** [comment]
you close this issue?

You really don't want users, seemingly.

**Comment by 5chdn** [comment]
Oh, I misunderstood, you said you are giving up?

**Comment by drandreaskrueger** [comment]
Yes. Because I do not know what else to try. Re-read last three comments please.

--> You run it. Show me what to do.

Thanks.

**Comment by drandreaskrueger** [comment]
run 14 = without the `--geth` compatibility switch  
https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-14

--> slower not faster. 

No further idea what to try.

**Comment by gituser** [comment]
@drandreaskrueger worth trying older version of parity (e.g. 1.8.x or 1.9.x or 1.10.x) and see if there is any difference at all. 

There was some change in code signing in 1.10 or in 1.9 (not sure though).

**Comment by drandreaskrueger** [comment]
Thanks for the idea, @gituser  

It might explain why I have seen faster rates with the Tobalaba fork (~ similiar to [parity version 1.8.0](https://gitlab.com/electronDLT/chainhammer/blob/master/tobalaba.md#ewfclient-fork-of-parity-180-)

---

And how?

In https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-14 it could be done here:

    sed -i 's/parity:stable/parity:v1.11.11/g' docker-compose.yml

    1.11.11 --> 1.8.x

BUT

```
curl -s 'https://registry.hub.docker.com/v2/repositories/parity/parity/tags/' | jq -r '."results"[]["name"]' | sort
beta
nightly
stable
v1.11.11
v2.0.4
v2.0.5
v2.0.5-rc0
v2.0.6
v2.1.0
v2.1.1
```
they have deleted all older versions.


**hello parity team - could you please re-create  the docker hub image of the most stable 1.8.x version?**

Thanks a lot.

**Comment by gituser** [comment]
>they have deleted all older versions.

weird, from github repository (just cloned freshly):

```
$ git clone https://github.com/paritytech/parity-ethereum
Cloning into 'parity-ethereum'...
cd pariremote: Enumerating objects: 52, done.
remote: Counting objects: 100% (52/52), done.
remote: Compressing objects: 100% (46/46), done.
Receiving objects:  21% (28972/137960), 16.01 MiB | 10.58 MiB/s    
Receiving objects:  60% (82776/137960), 40.14 MiB | 11.34 MiB/s     
remote: Total 137960 (delta 12), reused 14 (delta 6), pack-reused 137908
Receiving objects: 100% (137960/137960), 55.64 MiB | 9.40 MiB/s, done.
Resolving deltas: 100% (99246/99246), done.
Checking connectivity... done.
$ cd parity-ethereum
$ git tag -l
beta-0.9
beta-0.9.1
beta-release
mac-installer-hotfix
nightly
stable-release
test-tag
v1.0.0
v1.0.0-rc1
v1.0.1
v1.0.2
v1.1.0
v1.10.0
v1.10.0-ci0
v1.10.0-ci1
v1.10.0-ci2
v1.10.0-ci3
v1.10.0-ci4
v1.10.0-ci5
v1.10.0-ci6
v1.10.0-ci7
v1.10.1
v1.10.1-ci0
v1.10.2
v1.10.2-ci0
v1.10.2-ci1
v1.10.3
v1.10.3-ci0
v1.10.3-ci1
v1.10.4
v1.10.5
v1.10.5-ci0
v1.10.5-ci1
v1.10.5-rc0
v1.10.6
v1.10.7
v1.10.7-ci0
v1.10.8
v1.10.8-ci0
v1.10.8-ci1
v1.10.9
v1.10.9-rc0
v1.11.0
v1.11.0-ci0
v1.11.1
v1.11.10
v1.11.11
v1.11.2-ci0
v1.11.3
v1.11.4
v1.11.4-ci0
v1.11.5
v1.11.5-ci0
v1.11.5-ci1
v1.11.6
v1.11.6-rc0
v1.11.6-rc1
v1.11.6-rc2
v1.11.7
v1.11.7-rc0
v1.11.7-rc1
v1.11.7-rc2
v1.11.8
v1.12.0-ci0
v1.12.0-ci1
v1.12.0-ci2
v1.12.0-ci3
v1.12.0-ci4
v1.12.0-ci5
v1.2.0
v1.2.1
v1.2.2
v1.2.3
v1.2.4
v1.3.0
v1.3.1
v1.3.10
v1.3.11
v1.3.12
v1.3.13
v1.3.14
v1.3.15
v1.3.2
v1.3.3
v1.3.4
v1.3.5
v1.3.6
v1.3.7
v1.3.8
v1.3.9
v1.4.0
v1.4.1
v1.4.10
v1.4.11
v1.4.12
v1.4.2
v1.4.3
v1.4.4
v1.4.5
v1.4.6
v1.4.7
v1.4.8
v1.4.9
v1.5.0
v1.5.10
v1.5.11
v1.5.12
v1.5.13
v1.5.2
v1.5.3
v1.5.4
v1.5.6
v1.5.7
v1.5.8
v1.5.9
v1.6.0
v1.6.1
v1.6.10
v1.6.2
v1.6.3
v1.6.4
v1.6.5
v1.6.6
v1.6.7
v1.6.8
v1.6.9
v1.7.0
v1.7.1
v1.7.10
v1.7.11
v1.7.12
v1.7.13
v1.7.2
v1.7.3
v1.7.4
v1.7.5
v1.7.6
v1.7.7
v1.7.8
v1.7.9
v1.8.0
v1.8.1
v1.8.10
v1.8.10-ci0
v1.8.10-ci1
v1.8.10-ci2
v1.8.10-ci3
v1.8.10-ci4
v1.8.10-ci5
v1.8.11
v1.8.11-ci0
v1.8.2
v1.8.3
v1.8.4
v1.8.5
v1.8.6
v1.8.7
v1.8.8
v1.8.8-ci0
v1.8.8-ci1
v1.8.8-ci2
v1.8.8-ci3
v1.8.8-ci4
v1.8.9
v1.8.9-ci0
v1.9.0
v1.9.1
v1.9.1-ci0
v1.9.1-ci1
v1.9.1-ci2
v1.9.1-ci3
v1.9.2
v1.9.2-ci0
v1.9.3
v1.9.3-ci0
v1.9.3-ci1
v1.9.3-ci2
v1.9.3-ci3
v1.9.3-ci4
v1.9.3-ci5
v1.9.4
v1.9.4-ci0
v1.9.5
v1.9.5-ci0
v1.9.5-ci1
v1.9.5-ci2
v1.9.5-ci3
v1.9.5-ci4
v1.9.5-ci5
v1.9.5-ci6
v1.9.6
v1.9.6-ci0
v1.9.6-ci1
v1.9.6-ci2
v1.9.7
v1.9.7-ci0
v1.9.7-ci1
v2.0.0
v2.0.0-rc0
v2.0.0-rc1
v2.0.0-rc2
v2.0.0-rc3
v2.0.1
v2.0.3
v2.0.3-rc0
v2.0.4
v2.0.5
v2.0.5-rc0
v2.0.6
v2.1.0
v2.1.0-rc0
v2.1.0-rc1
v2.1.0-rc2
v2.1.0-rc3
v2.1.0-rc4
v2.1.1
v2.2.0-rc0
```

**Comment by drandreaskrueger** [comment]
I would probably try `v1.7.13`, to be on the safe side. 

Did v1.7 already have aura?

Did it have `instaseal` developmentChain, so would `parity-deploy.sh --dev` work ? That already would show something probably.

**Comment by 5chdn** [comment]
> hello parity team - could you please re-create the docker hub image of the most stable 1.8.x version?

we don't delete anything. just checkout the tag directly, either on github or on docker

> Did v1.7 already have aura?

yes.

**Comment by drandreaskrueger** [comment]
yes.

thanks.

**Comment by drandreaskrueger** [comment]
hello gituser, Re: [your comment above](https://github.com/paritytech/parity-ethereum/issues/9393#issuecomment-423364834)  

> @gituser  
> gituser commented 4 days ago •   
> @drandreaskrueger worth trying older version of parity (e.g. 1.8.x or 1.9.x or 1.10.x)   
> and see if there is any difference at all.  
> There was some change in code signing in 1.10 or in 1.9 (not sure though).  

I had a lot of hope when you said that.

But I have tried some older versions now:

(run15) v1.7.13 and instantseal https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-15  
(run16) v1.7.13 and aura https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-16  
(run17) v1.8.11 and aura https://gitlab.com/electronDLT/chainhammer/blob/master/parity.md#run-17  

not faster.

**Comment by 5chdn** [comment]
> not faster.

would be surprised if the speed varies across versions

**Comment by drandreaskrueger** [comment]
> would be surprised if the speed varies across versions

yes me too.



but in lack of any other substantial suggestions, and as a test of his comment

```
@gituser
gituser commented 4 days ago •
@drandreaskrueger worth trying older version of parity (e.g. 1.8.x or 1.9.x or 1.10.x)
and see if there is any difference at all.
There was some change in code signing in 1.10 or in 1.9 (not sure though).
```

I had to try that.

**Comment by drandreaskrueger** [comment]
I could get contact with CodyBorn from Microsoft, he answered to my tweets.

I have summarized what he is revealing about his approach here: https://gitlab.com/electronDLT/chainhammer/blob/master/codyborn.md

**Comment by drandreaskrueger** [comment]
> CodyBorn

What is clear already: He is "too far out" to be applicable to my simple benchmarking. 

I don't feel like creating hundreds of sender-accounts just to bypass nonce lookup and then sign my own transactions. 

And even if that made parity faster, for me it would simply mean that paritytech should revisit that part of your parity code, to accelerate it - and repeatedly re-run my chainhammer, to notice when you got it faster. 

Because his approach with `web3.eth.sendRawTransaction()` would not be practical for our daily use of `parity`; we want the Ethereum node to do that work for us efficiently.

### probably the most important hint for paritytech

it is probably something outside of sendRawTransaction() but within the parity code base which is slowing down transactions by a factor of more than 500%

### the most important experiment that CodyBorn could do:  

replicate his exact approach but instead of `pairy aura` using `geth clique`, see [geth.md#results](https://gitlab.com/electronDLT/chainhammer/blob/master/geth.md#results).

**Comment by drandreaskrueger** [comment]
parity v2.2.0 (single-threaded) seems slightly faster than v1.11.11 (multi-threaded), see these brand new results: https://github.com/paritytech/parity-ethereum/issues/9582#issuecomment-427144992

**Comment by drandreaskrueger** [comment]
Hey @5chdn @Tbaut @ddorgan @AyushyaChitransh @gituser 

Greetings from Berlin, web3summit.

Actually, **who of you** are you **in Berlin** now? We should meet up this week!

**Comment by ddorgan** [comment]
So I just did a test on a c5.xlarge but only using the --geth extra option and I'm seeing this:

```
blocknumber_start_here = 17
starting timer, at block 17 which has  1  transactions; at timecode 43450.223308099
block 17 | new #TX 661 / 4000 ms = 165.2 TPS_current | total: #TX  662 /  4.0 s = 166.9 TPS_average
block 18 | new #TX 775 / 4000 ms = 193.8 TPS_current | total: #TX 1437 /  8.0 s = 179.9 TPS_average
block 19 | new #TX 750 / 4000 ms = 187.5 TPS_current | total: #TX 2187 / 12.0 s = 182.4 TPS_average
block 20 | new #TX 763 / 4000 ms = 190.8 TPS_current | total: #TX 2950 / 16.0 s = 184.3 TPS_average
block 21 | new #TX 740 / 4000 ms = 185.0 TPS_current | total: #TX 3690 / 20.0 s = 184.4 TPS_average
block 22 | new #TX 740 / 4000 ms = 185.0 TPS_current | total: #TX 4430 / 24.0 s = 184.4 TPS_average
block 23 | new #TX 758 / 4000 ms = 189.5 TPS_curr, wient | total: #TX 5188 / 28.3 s = 183.2 TPS_average
block 24 | new #TX 717 / 4000 ms = 179.2 TPS_current | total: #TX 5905 / 32.0 s = 184.5 TPS_average
block 25 | new #TX 751 / 4000 ms = 187.8 TPS_current | total: #TX 6656 / 36.0 s = 184.8 TPS_average
block 26 | new #TX 712 / 4000 ms = 178.0 TPS_current | total: #TX 7368 / 40.0 s = 184.2 TPS_average
block 27 | new #TX 733 / 4000 ms = 183.2 TPS_current | total: #TX 8101 / 44.0 s = 184.1 TPS_average
block 28 | new #TX 728 / 4000 ms = 182.0 TPS_current | total: #TX 8829 / 48.0 s = 183.9 TPS_average
block 29 | new #TX 742 / 4000 ms = 185.5 TPS_current | total: #TX 9571 / 52.0 s = 183.9 TPS_average
```

Still seems quite single thread heavy though, will try with some options to speed that up.

**Comment by drandreaskrueger** [comment]
Great results. Looks very promising.

We have to find out how to prevent this https://github.com/paritytech/parity-ethereum/issues/9582 to always get to those results, consistently. Please let it run to the 20k end - perhaps even with more than 20k transactions, see config.py

And: Keep me in the loop whatever you find out together with Tomek.  

Thanks again for your time on Thursday in the parity office. Great working with you!

Greetings from the train to Prague.

**Comment by drandreaskrueger** [comment]
just found this:

> ... validation logic, can be separated into three sections:  
> • Validation that actions are internally consistent  
> • Validation that preconditions are met  
> • Modification of the application state  
> The read only aspects of this process can be conducted  
> in parallel, while modification which requires write access  
> must be conducted sequentially ...

is that also how parity is internally digesting multi-threaded transaction requests?

**Comment by ddorgan** [comment]
@drandreaskrueger let me come back to you on this.

The bottleneck is most probably the signing itself. I think for geth you are basically pre-signing them because the web3 library is too slow, right? 

I may make a change to your script to do the same for parity so that transactions are signed before being submitted. 

This would like up with the geth process, right?

**Comment by drandreaskrueger** [comment]
> because the web3 library is too slow

Yes, for low (two digit) TPS it does not make a big difference, only 20% faster. But when I get into the hundreds of TPS, I see considerable gains (twice as fast) when bypassing web3 completely.  Please have a quick look at these old experiments: https://github.com/drandreaskrueger/chainhammer/blob/master/log.md#sending-via-web3-versus-sending-via-rpc

---


> basically pre-signing 

Not sure we are talking about the same thing actually. Even when bypassing the web3.py library, I am using the RPC `method = 'eth_sendTransaction`.


Have a look at these two codepieces:

### via web3  

in https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/send.py#L73-L93   
it is simply this one liner https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/send.py#L90

while 

### via RPC

in https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/send.py#L106-L183
I am doing (`contract_method_ID` + `argument_encoding` --> `payload`), then (plus `headers` into a`requests.post` to call `eth_sendTransaction`), see here:  
https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/send.py#L178


### choice

I switch between those two routes here https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/send.py#L201

choice constant `ROUTE` is defined in config.py

---

>  to do the same for parity so that transactions are signed before being submitted. This would like up with the geth process, right?

No, no difference between the two.

As long as `PARITY_UNLOCK_EACH_TRANSACTION=False` https://github.com/drandreaskrueger/chainhammer/blob/93c40384a4d178bdb00cea491d15b14046471b72/config.py#L43 is False, chainhammer is talking to geth identically how it is talking to parity.

**Comment by drandreaskrueger** [comment]
see above.

> The bottleneck is most probably the signing itself.

I suggest you compare the tx signing part of the geth code with the parity code.

**Comment by drandreaskrueger** [comment]
The newest **chainhammer v55** is fully automated - you run & analyze _a whole experiment with one or two CLI lines_. I am optimistic that you find a clever combination of parity CLI switches that will speed it up. Good luck. 

Because of general interest, I have created this new issue:
https://github.com/paritytech/parity-ethereum/issues/10382

**Comment by 5chdn** [comment]
Thanks for sharing :+1:

**Comment by drandreaskrueger** [comment]
> closed this

So you want to track the speed improvements in the new issue. Yes, that makes sense.
